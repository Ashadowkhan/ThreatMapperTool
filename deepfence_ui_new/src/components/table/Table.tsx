import {
  Cell,
  ColumnDef,
  createColumnHelper,
  flexRender,
  getCoreRowModel,
  getPaginationRowModel,
  Header,
  HeaderGroup,
  OnChangeFn,
  PaginationState,
  Row,
  RowData,
  RowModel,
  useReactTable,
} from '@tanstack/react-table';
import cx from 'classnames';
import { once } from 'lodash-es';
import { createContext, Fragment, useContext } from 'react';
import { FaMinus, FaPlus } from 'react-icons/fa';

import IconButton from '../button/IconButton';
import Pagination from '../pagination/Pagination';
import { Typography } from '../typography/Typography';

export interface TableProps<TData extends RowData> {
  data: TData[];
  columns: ColumnDef<TData, any>[];
  renderSubComponent?: (props: { row: Row<TData> }) => React.ReactElement;
  getRowCanExpand?: (row: Row<TData>) => boolean;
  striped?: boolean;
  enablePagination?: boolean;
  manualPagination?: boolean;
  pageIndex?: number;
  pageSize?: number;
  pageCount?: number;
  onPaginationChange?: OnChangeFn<PaginationState>;
}

interface TableContextValues<TData extends RowData> {
  striped?: boolean;
  renderSubComponent?: (props: { row: Row<TData> }) => React.ReactElement;
}

const createTableContext = once(<TData extends RowData>() =>
  createContext<TableContextValues<TData>>({}),
);
function useTableContext<TData extends RowData>() {
  return useContext(createTableContext<TData>());
}

export function Table<TData extends RowData>(props: TableProps<TData>) {
  const {
    data,
    columns,
    striped,
    renderSubComponent,
    getRowCanExpand,
    enablePagination,
    manualPagination,
    pageIndex = 0,
    pageSize = 10,
    pageCount = -1,
    onPaginationChange,
  } = props;
  const TableContext = createTableContext<TData>();
  const table = useReactTable<TData>({
    data,
    columns,
    getCoreRowModel: getCoreRowModel(),
    getPaginationRowModel: getPaginationRowModel(),
    getRowCanExpand,
    columnResizeMode: 'onChange',
    meta: {
      striped,
    },
    state: {
      ...(manualPagination
        ? {
            pagination: {
              pageIndex,
              pageSize,
            },
          }
        : {}),
    },
    ...(manualPagination
      ? { manualPagination: true, onPaginationChange, pageCount }
      : {}),
  });

  const [headerGroups, rowModel] = [table.getHeaderGroups(), table.getRowModel()];

  return (
    <TableContext.Provider value={{ striped, renderSubComponent }}>
      <div
        className={cx(
          `overflow-hidden`,
          `shadow-[0px_1px_3px_rgba(0,_0,_0,_0.1),_0px_1px_2px_-1px_rgba(0,_0,_0,_0.1)] dark:shadow-sm`,
          `rounded-lg dark:border dark:border-gray-700`,
        )}
      >
        <table
          className={cx(
            `w-full bg-white dark:bg-gray-800 border-spacing-0 border-collapse`,
          )}
          cellPadding="0"
          cellSpacing="0"
        >
          <TableHead headerGroups={headerGroups} />
          <TableBody rowModel={rowModel} />
        </table>
      </div>
      {enablePagination ? (
        <div className="mt-4 flex justify-end">
          <Pagination
            currentPage={table.getState().pagination.pageIndex + 1}
            onPageChange={(page) => {
              table.setPageIndex(page - 1);
            }}
            totalPageCount={table.getPageCount()}
          />
        </div>
      ) : null}
    </TableContext.Provider>
  );
}

function TableHead<TData>({ headerGroups }: { headerGroups: HeaderGroup<TData>[] }) {
  return (
    <thead className="bg-gray-50 dark:bg-gray-700">
      {headerGroups.map((headerGroup) => (
        <tr key={headerGroup.id}>
          {headerGroup.headers.map((header) => (
            <Th header={header} key={header.id} />
          ))}
        </tr>
      ))}
    </thead>
  );
}

function Th<TData>({ header }: { header: Header<TData, unknown> }) {
  return (
    <th
      key={header.id}
      colSpan={header.colSpan}
      className={cx(
        'relative border-0 text-gray-500 dark:text-white p-4 text-left',
        'border-b border-gray-200 dark:border-gray-700',
        Typography.size.xs,
        Typography.weight.semibold,
        Typography.decoration.uppercase,
      )}
      style={{ width: header.getSize() }}
    >
      {header.isPlaceholder
        ? null
        : flexRender(header.column.columnDef.header, header.getContext())}
      {false && header.column.getCanResize() && (
        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
        <div
          onMouseDown={header.getResizeHandler()}
          onTouchStart={header.getResizeHandler()}
          className={`absolute right-0 top-0 h-full w-1 bg-gray-500 cursor-col-resize select-none`}
        ></div>
      )}
    </th>
  );
}

function TableBody<TData>({ rowModel }: { rowModel: RowModel<TData> }) {
  const { striped, renderSubComponent } = useTableContext<TData>();
  return (
    <tbody>
      {rowModel.rows.map((row) => (
        <Fragment key={row.id}>
          <tr
            className={cx(
              {
                'bg-gray-50 dark:bg-gray-700': striped && row.index % 2 !== 0,
              },
              `hover:!bg-gray-100 dark:hover:!bg-gray-600`,
              'transition-colors',
            )}
          >
            {row.getVisibleCells().map((cell) => (
              <Td cell={cell} key={cell.id} totalRows={rowModel.rows.length} />
            ))}
          </tr>
          {row.getIsExpanded() && (
            <tr>
              <td
                colSpan={row.getVisibleCells().length}
                className="border-b border-t border-gray-200 dark:border-gray-700"
              >
                {renderSubComponent?.({ row })}
              </td>
            </tr>
          )}
        </Fragment>
      ))}
    </tbody>
  );
}

function Td<TData>({
  cell,
  totalRows,
}: {
  cell: Cell<TData, unknown>;
  totalRows: number;
}) {
  const { striped } = useTableContext<TData>();

  return (
    <td
      key={cell.id}
      style={{ width: cell.column.getSize() }}
      className={cx(
        `p-4 text-sm text-gray-900 dark:text-white`,
        Typography.weight.normal,
        {
          'border-b border-gray-200 dark:border-gray-700':
            !striped && cell.row.index !== totalRows - 1,
        },
      )}
    >
      {flexRender(cell.column.columnDef.cell, cell.getContext())}
    </td>
  );
}

export function RowExpander<TData extends RowData>({ row }: { row: Row<TData> }) {
  return row.getCanExpand() ? (
    <IconButton
      color="primary"
      icon={row.getIsExpanded() ? <FaMinus /> : <FaPlus />}
      onClick={row.getToggleExpandedHandler()}
      outline
      size="xs"
    />
  ) : null;
}

export { createColumnHelper };
