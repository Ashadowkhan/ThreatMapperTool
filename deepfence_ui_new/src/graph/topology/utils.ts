/**
 * Topology Graph utilities
 *
 * How Topology graph is working?
 * 1. To render a graph we need nodes and edges
 * 2. Backend sends these nodes and edges as delta, frontend transform these data into g6 structure with additional
 * attributes store in each node model which could be either simple node or combo node. The attributes help in node expansion and collapsion
 * 3. Each time a new node is received or old node is deleted backend sends only those information. Graph dynamically
 * update graph.
 * 4. Topology uses GForce layout to put corresponding nodes close to each other with an attempt to put meaningful nodes in the
 * center of the screen
 * 5. For combo nodes, the nodes are group by a circle and all nodes inside it are attracted towards center of circle. This is achieved with
 * the help of a node which is invisible to user and it helps to pull all of those nodes otherwise random nodes may appear on top of combo circle.
 * 6. Graph has node, edges, children nodes and edges. Each time a new data is received, we need to update them. This take place in such way:
 *   a) remove all edges mentioned in api data
 *   b) removes all nodes including child nodes and edges mentioned in api data and add new nodes in case of node expansion
 *   c) add edges to newly connected nodes
 *
 */

import { ICombo, IGraph, Item } from '@antv/g6';

import { StringIndexType } from '../../topology/topology-client';
import {
  collapseNode,
  finishExpandingNode,
  itemExpandsAsCombo,
  itemIsExpanded,
  itemIsExpanding,
} from '../graphManager/expand-collapse';
import { pointAround } from '../graphManager/gforce';
// import { IG6GraphEvent, IUserNode } from '@antv/graphin';
import { COLORS, PALETTE } from '../theme';
import { ApiNodeItemType, ICustomNode, IItem, INode, IStringIndex } from '../types';

export interface IAPIData {
  nodes: {
    add: ApiNodeItemType[];
    update: ApiNodeItemType[];
    remove: string[];
    reset?: boolean;
  };
  edges: {
    add: ApiNodeItemType[];
    update: ApiNodeItemType[];
    remove: ApiNodeItemType[]; // edge remove is not string array in api response though node remove is
    reset?: boolean;
  };
  reset: boolean;
  metadata: ApiNodeItemType;
}

export const nodeSize = (node_type: string) => {
  const mul: StringIndexType<number> = {
    container: 0.5,
    process: 0.5,
  };

  const size = (mul[node_type] || 1) * 60;
  return size;
};

export const modelParentsToTopologyParents = (nodes: ICustomNode[]) => {
  const ret: IStringIndex<string> = {};

  for (const node of nodes) {
    const node_type = uiToServerNodeMap(node.node_type);
    if (node_type !== null) {
      ret[node_type] = node.id;
    }
  }

  return ret;
};

export const modelNodeTypeToTopologyChildrenTypes = (
  type: string,
  opts: {
    [k: string]: boolean;
  },
) => {
  const types: { [k: string]: string[] } = {
    cloud: [TopologyNodeType.REGION, TopologyNodeType.KUBERNETES_CLUSTER],
    region: [TopologyNodeType.HOST],
    kubernetes_cluster: [TopologyNodeType.HOST],
    host: opts?.kubernetes
      ? [TopologyNodeType.POD]
      : [TopologyNodeType.PROCESS, TopologyNodeType.CONTAINER],
    pod: [TopologyNodeType.CONTAINER],
    container: [TopologyNodeType.PROCESS],
  };

  return types[type];
};

export const uiToServerNodeParents = (nodes: ICustomNode[]) => {
  const ret: { [key: string]: string } = {};

  for (const node of nodes) {
    const node_type = uiToServerNodeMap(node.node_type);
    if (node_type) {
      ret[node_type] = node.id;
    }
  }

  return ret;
};

export const TopologyNodeType = {
  CLOUD_PROVIDER: 'cloud-providers',
  REGION: 'cloud-regions',
  KUBERNETES_CLUSTER: 'kubernetes-clusters',
  HOST: 'hosts',
  POD: 'pods',
  CONTAINER: 'containers',
  PROCESS: 'processes',
};

export const serverToUINodeMap = (type: string): string => {
  const vals: { [key: string]: string } = {
    '<cloud_provider>': 'cloud',
    '<cloud_region>': 'region',
    '<kubernetes_cluster>': 'kubernetes_cluster',
    '<host>': 'host',
    '<pod>': 'pod',
    '<container>': 'container',
    '<fargate>': 'fargate',
  };

  return vals[type];
};

export const uiToServerNodeMap = (type: string) => {
  const types: { [key: string]: string } = {
    cloud: TopologyNodeType.CLOUD_PROVIDER,
    region: TopologyNodeType.REGION,
    kubernetes_cluster: TopologyNodeType.KUBERNETES_CLUSTER,
    host: TopologyNodeType.HOST,
    pod: TopologyNodeType.POD,
    container: TopologyNodeType.CONTAINER,
    process: TopologyNodeType.PROCESS,
  };

  const ret = types[type];
  if (ret === undefined) {
    console.error('no topology type for model node type', type);
    return null;
  }

  return ret;
};

export const getConditionalFontSize = (node_type: string) => {
  switch (node_type) {
    case 'pod':
    case 'container':
    case 'process':
      return 14;
  }
};

// collapse or expand cases
export const itemExpands = (item: IItem) => {
  const model = item?.get?.('model');

  switch (model.node_type) {
    case 'cloud':
    case 'region':
    case 'kubernetes_cluster':
    case 'host':
    case 'pod':
    case 'container':
      return true;
  }

  return false;
};

export const pseudoEdge = (source: string, target: string) => ({
  source,
  target,
  pseudo: true,
});

const CLOUD_STYLES = [PALETTE.GOOGLE_BLUE, PALETTE.AWS_YELLOW];
const cloudInfo = () => {
  const color = CLOUD_STYLES.shift() as string;
  CLOUD_STYLES.push(color);
  return {
    nodeStyle: {
      fill: color || COLORS.NODE,
    },
    edgeStyle: {
      stroke: COLORS.EDGE,
    },
  };
};

const removeNodesCombo = (graph: IGraph, item: Item, nodes: string[]) => {
  const model = item.get('model');
  const node_id: string = model.id;
  const combo_id = `${node_id}-combo`;
  const combo_model = graph.findById(combo_id).get('model');
  const combo_children_ids = combo_model.children_ids;

  for (const node_id of nodes) {
    if (!combo_children_ids.has(node_id)) {
      console.error('trying to remove unknown child from combo', combo_id, node_id);
      continue;
    }
    combo_children_ids.delete(node_id);

    const item = graph.findById(node_id) as ICustomNode;
    if (itemIsExpanded(item)) {
      collapseNode(graph, item);
    }
    removeNodeItem(graph, item);
  }
};

const addNodesCombo = (graph: IGraph, item: IItem, nodes: ApiNodeItemType[]) => {
  const model = item.get('model');
  const node_id = model.id;
  const combo_id = `${node_id}-combo`;

  const combo = graph.findById(combo_id);
  const combo_model = combo.get('model');
  const children_ids = combo_model.children_ids;

  const center_id = combo_model.center_ids[0];
  const center_model = graph.findById(center_id).get('model');

  const n_nodes = children_ids.size + nodes.length;

  const cloudInfo = model.cloudInfo;
  for (const node of nodes) {
    graph.addItem('node', {
      ...node,
      style: {
        ...nodeStyle(node, cloudInfo?.nodeStyle),
      },
      cloudInfo,
      parent_id: node_id,
      comboId: combo_id,
      children_ids: new Set([]),
      x: n_nodes > 1 ? pointAround(center_model.x) : center_model.x,
      y: n_nodes > 1 ? pointAround(center_model.y) : center_model.y,
    });

    graph.addItem('edge', {
      ...pseudoEdge(center_id, node.id),
      combo_pseudo_inner: true,
      style: { lineWidth: 0, endArrow: false },
    });
    children_ids.add(node.id);
  }
};

export const updateComboNode = (graph: IGraph, item: IItem, delta: IAPIData['nodes']) => {
  if (delta.remove) {
    removeNodesCombo(graph, item, delta.remove);
  }

  if (delta.add) {
    addNodesCombo(graph, item, delta.add);
  }
};

const removeNodesSimple = (graph: IGraph, item: Item, nodes: string[]) => {
  const model = item.get('model');
  const children_ids = model.children_ids;

  for (const child_node_id of nodes) {
    const child = graph.findById(child_node_id) as ICustomNode;
    if (!child || !children_ids.has(child_node_id)) {
      console.error('trying to remove an unknown child', child_node_id);
      continue;
    }
    children_ids.delete(child_node_id);

    if (itemIsExpanded(child)) {
      collapseNode(graph, child, undefined, false);
    }
    removeNodeItem(graph, child);
  }
};

const addNodesSimple = (graph: IGraph, item: Item, nodes: ApiNodeItemType[]) => {
  const model = item.get('model');
  const node_id = model.id;
  const children_ids = model.children_ids;
  const cloudInfo = model.cloudInfo;

  for (const node of nodes) {
    const item = graph.findById(node.id);
    if (item !== undefined) {
      console.error(
        `trying to add node that is already in the graph (parent=${node_id})`,
        node,
      );
      continue;
    }

    const node_item = graph.addItem('node', {
      ...node,
      x: pointAround(model.x),
      y: pointAround(model.y),
      parent_id: node_id,
      cloudInfo,
      style: { ...nodeStyle(node, cloudInfo?.nodeStyle) },
      children_ids: new Set(),
    }) as Item;
    graph.addItem('edge', {
      ...pseudoEdge(node_id, node.id),
      style: { ...model.cloudInfo?.edgeStyle },
    });
    children_ids.add(node.id);
    node_item.refresh();
  }
};

export const updateSimpleNode = (graph: IGraph, item: Item, delta: IAPIData['nodes']) => {
  if (delta.remove) {
    removeNodesSimple(graph, item, delta.remove as string[]);
  }

  if (delta.add) {
    addNodesSimple(graph, item, delta.add);
  }
};

export const updateGraphNode = (graph: IGraph, item: IItem, delta: IAPIData['nodes']) => {
  if (itemIsExpanding(item)) {
    finishExpandingNode(graph, item);
  }

  if (itemExpandsAsCombo(item)) {
    return updateComboNode(graph, item, delta);
  }

  return updateSimpleNode(graph, item, delta);
};

export const removeNodeItem = (graph: IGraph, item: INode) => {
  for (const edge of item.getEdges()) {
    const edge_model = edge.get('model');
    if (edge_model.connection) {
      graph.removeItem(edge_model.id);
    }
  }

  graph.removeItem(item);
};

const nodeStyle = (node: StringIndexType<any>, override: StringIndexType<any>) => {
  let style: StringIndexType<string> = {};
  const fill: IStringIndex<string> = {
    cloud: COLORS.CLOUD_PROVIDER,
    region: COLORS.REGION,
    host: COLORS.HOST,
    pod: COLORS.POD,
    container: COLORS.CONTAINER,
    process: COLORS.PROCESS,
  };
  style.fill = fill[node.node_type] || COLORS.NODE;

  style = { ...style, ...override };
  if (node.img !== undefined) {
    delete style.fill;
  } else if (node.node_type === 'process') {
    style.fill = COLORS.PROCESS;
  }

  return style;
};

export const updateGraphRootNodes = (graph: IGraph, delta: StringIndexType<any>) => {
  for (const node_id of delta.remove || []) {
    const node = graph.findById(node_id) as INode;
    if (node === undefined) {
      console.error('trying to remove unknown root node', node_id);
      continue;
    }
    removeNodeItem(graph, node);
  }

  const center_x = graph.getWidth() / 2;
  const center_y = graph.getHeight() / 2;

  for (const node of delta.add || []) {
    const info = node.node_type === 'cloud' ? cloudInfo() : null;

    graph.addItem('node', {
      ...node,
      x: pointAround(center_x),
      y: pointAround(center_y),
      cloudInfo: info,
      style: nodeStyle(node, info?.nodeStyle ?? {}),
      children_ids: new Set(),
    });
  }
};

export const updateGraphEdges = (graph: IGraph, delta: IAPIData['edges']) => {
  const removeEdge = (item: IItem) => {
    const model = item.get('model');
    if (model.connection === true) {
      graph.removeItem(model.id);
    }
  };

  if (delta.reset) {
    for (const edge of graph.getEdges()) {
      removeEdge(edge);
    }
  }

  if (delta.add) {
    const r = [];
    for (const edge of delta.add) {
      const source = graph.findById(edge.source)?.get('model');
      if (source === undefined) {
        console.error('edge source does not exist', edge);
        continue;
      }
      const target = graph.findById(edge.target)?.get('model');
      if (target === undefined) {
        console.error('edge target does not exist', edge);
        continue;
      }

      graph.addItem('edge', {
        ...edge,
        style: source.cloudInfo?.edgeStyle,
        connection: true,
      });
      r.push({
        ...edge,
        style: source.cloudInfo?.edgeStyle,
        connection: true,
      });
    }
    return r;
  }

  if (delta.remove) {
    for (const edge of delta.remove) {
      const id = `${edge.source}-${edge.target}`;
      const item = graph.findById(id);
      if (item === undefined) {
        console.warn("trying to remove edge that doesn't exist", edge.id);
        continue;
      }

      removeEdge(item);
    }
  }
};

export const getParents = (graph: IGraph, item: Item) => {
  const parents: string[] = [];

  let parent_id: string = item.get('model').parent_id;
  while (parent_id) {
    let parent: ICustomNode = graph.findById(parent_id) as ICustomNode;
    if (parent) {
      parent = parent.get<ICustomNode>('model');
      if (parent.node_type !== 'combo') {
        parents.unshift(parent_id);
      }
    }

    parent_id = parent.parent_id;
  }

  return parents;
};

export const onHoverNode = (item: ICustomNode, hover: boolean) => {
  const model = item.get('model');
  if (model.node_type === 'process') {
    if (hover) {
      item.update({ label: model.label_full });
      item.toFront();
    } else {
      item.update({ label: model.label_short });
    }
  }
};

export const fixCombo = (graph: IGraph, combo: ICombo) => {
  const model = combo.get('model');
  const bbox = combo.getBBox();
  const { centerX, centerY } = bbox;

  const center_id = model.center_ids[0];
  const center = graph.findById(center_id);
  const center_model = center.get('model');

  center_model.fx = centerX;
  center_model.fy = centerY;
};
