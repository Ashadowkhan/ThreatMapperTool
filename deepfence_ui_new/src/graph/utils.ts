/**
 * General Graph utilities
 */

import {
  Graph,
  GraphinData,
  IG6GraphEvent,
  IUserEdge,
  IUserNode,
  Utils,
} from '@antv/graphin';
import { Combo } from '@antv/graphin/lib/typings/type';

type GraphinItemType = IUserNode | IUserEdge | Combo[] | undefined | null;
export type GraphItem = IG6GraphEvent['item'];

// TODOï¼šbuild-in Graphin.Utils
export const update = (data: GraphinData, type: 'node' | 'edge' = 'node') => {
  const items: GraphinItemType[] = data[`${type}s`];
  return {
    set: (id: string, model: any) => {
      const newItems: GraphinItemType[] = [];
      items.forEach((item: GraphinItemType) => {
        if ((item as IUserNode | IUserEdge)?.id === id) {
          const mergedItem = Utils.deepMix({}, item, model);
          newItems.push(mergedItem);
        } else {
          newItems.push(item);
        }
      });
      return {
        ...data,
        [`${type}s`]: newItems,
      };
    },
  };
};

export function arrayTransformByFunction<T>(
  arrays: T[],
  fn: (element: T) => T | null,
): T[] {
  return arrays.reduce((acc: T[], element: T) => {
    const result = fn(element);
    if (result) {
      acc.push(result);
    }
    return acc;
  }, []);
}

export const basename = (path: string) => {
  const i = path.lastIndexOf('/');
  if (i >= 0) {
    return path.substring(i + 1);
  }
  return path;
};

export const ellipsize = (text: string, n: number) => {
  if (text.length <= n) {
    return text;
  }

  return text.substring(0, n - 3) + '...';
};

export const itemSetExpanded = (item) => {
  const model = item.get('model');
  model.expand_state = ExpandState.EXPANDED;
};

export const itemIsExpanded = (item: GraphItem) => {
  return item?.get?.('model')?.expand_state !== undefined;
};

const itemUnsetExpanded = (item: GraphItem) => {
  const model = item?.get?.('model');
  delete model.expand_state;
};

export const removeNodeItem = (graph: Graph, item: GraphItem) => {
  if (!item) {
    return;
  }
  for (const edge of item.getEdges()) {
    const edge_model = edge.get('model');
    if (edge_model.connection) {
      graph.removeItem(edge_model.id);
    }
  }

  const model = item?.get?.('model');
  console.log('removing node', model.id);
  graph.removeItem(item);
};

export const collapseSimpleNode = (
  graph: Graph,
  item: GraphItem,
  onNodeCollapsed: any,
  isChild: boolean,
) => {
  const model = item?.get?.('model');
  const node_id = model.id;

  const edges = item?.getOutEdges?.() ?? [];
  for (const edge of edges) {
    graph.removeItem(edge.id);
  }

  if (model.children_ids) {
    for (const child_id of model.children_ids) {
      const child = graph.findById(child_id);

      if (itemIsExpanded(child)) {
        collapseSimpleNode(graph, child, onNodeCollapsed, true);
      }
      removeNodeItem(graph, child);
    }

    model.children_ids.clear();
  }

  itemUnsetExpanded(item);

  if (onNodeCollapsed) {
    onNodeCollapsed(item, isChild);
  }
};
