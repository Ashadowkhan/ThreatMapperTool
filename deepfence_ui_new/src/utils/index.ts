export function basePath(urlPath: string) {
  //
  // "/scope/terminal.html" -> "/scope"
  // "/scope/" -> "/scope"
  // "/scope" -> "/scope"
  // "/" -> ""
  //
  const parts = urlPath.split('/');
  // if the last item has a "." in it, e.g. foo.html...
  if (parts[parts.length - 1].indexOf('.') !== -1) {
    return parts.slice(0, -1).join('/');
  }
  return parts.join('/').replace(/\/$/, '');
}

export function getWebsocketUrl(
  host = window.location.host,
  pathname = window.location.pathname,
) {
  //   if (process.env.WS_BASE_URL) {
  //     return process.env.WS_BASE_URL;
  //   }
  return 'wss://threatmapper.deepfence.show';
  const wsProto = window.location.protocol === 'https:' ? 'wss' : 'ws';
  return `${wsProto}://${host}${process.env.SCOPE_API_PREFIX || ''}${basePath(pathname)}`;
}
