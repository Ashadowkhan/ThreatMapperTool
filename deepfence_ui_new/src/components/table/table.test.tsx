import '@testing-library/jest-dom';

import { fireEvent, waitFor } from '@testing-library/react';

import { renderWithClient } from '../../tests/utils';
import { getRowExpanderColumn, Table, TableProps } from './Table';

interface Fruit {
  id: number;
  name: string;
  description: string;
}

function createDummyFruitData(length = 2) {
  const data: Fruit[] = [];
  for (let i = 0; i < length; i++) {
    data.push({
      id: i,
      name: `Fruit ${i}`,
      description: `Description for ${i}`,
    });
  }
  return data;
}

const defaultColumns = [
  {
    accessorKey: 'id',
  },
  {
    accessorKey: 'name',
  },
  {
    accessorKey: 'description',
  },
];

const BasicTable = ({
  dataLength,
  tableProps,
}: {
  dataLength?: number;
  tableProps?: Partial<TableProps<Fruit>>;
}) => {
  return (
    <Table
      data={createDummyFruitData(dataLength)}
      columns={defaultColumns}
      {...tableProps}
    />
  );
};

describe(`Component Table`, () => {
  it('should render a basic table', async () => {
    const { getByRole, getAllByTestId, getAllByRole } = renderWithClient(<BasicTable />);
    expect(getByRole('table')).toBeInTheDocument();
    expect(getAllByTestId('table-header-row').length).toEqual(1);
    expect(getAllByRole('columnheader').length).toEqual(3);
    const rows = getAllByRole('row');
    expect(rows.length).toEqual(3);
    const cells = getAllByRole('cell');
    expect(cells.length).toEqual(6);
    cells.forEach((cell, index) => {
      if (index < 3) {
        expect(cell.outerHTML).toMatchSnapshot(`cells with border ${index}`);
      } else {
        expect(cell.outerHTML).toMatchSnapshot(`cells without border ${index}`);
      }
    });
    fireEvent.mouseOver(rows[2]);

    await waitFor(() => {
      expect(rows[2]).toMatchSnapshot('row with hover state on normal table');
    });
  });

  it('should render a striped table', async () => {
    const { getByRole, getAllByTestId, getAllByRole } = renderWithClient(
      <BasicTable
        tableProps={{
          striped: true,
        }}
      />,
    );
    expect(getByRole('table')).toBeInTheDocument();
    expect(getAllByTestId('table-header-row').length).toEqual(1);
    expect(getAllByRole('columnheader').length).toEqual(3);
    const rows = getAllByRole('row');
    expect(rows.length).toEqual(3);
    const cells = getAllByRole('cell');
    expect(cells.length).toEqual(6);
    expect(rows[1]).toMatchSnapshot('odd rows should have lighter background');
    expect(rows[2]).toMatchSnapshot('even rows should have darker background');
    expect(cells[1]).toMatchSnapshot(`cells should not have border`);

    fireEvent.mouseOver(rows[2]);

    await waitFor(() => {
      expect(rows[2]).toMatchSnapshot('row with hover state on striped table');
    });
  });

  it('expandable rows should expand correctly', async () => {
    const { getAllByRole, getByText, queryByText } = renderWithClient(
      <BasicTable
        dataLength={5}
        tableProps={{
          renderSubComponent: ({ row }) => {
            return <div>row id {row.original.id}</div>;
          },
          getRowCanExpand: (row) => {
            return row.index !== 0;
          },
          columns: [getRowExpanderColumn<Fruit>(), ...defaultColumns],
        }}
      />,
    );
    const expanderButtons = getAllByRole('button');
    expect(expanderButtons.length).toEqual(4);

    expect(queryByText('row id 2')).not.toBeInTheDocument();

    fireEvent.click(expanderButtons[1]);

    await waitFor(() => {
      expect(getByText('row id 2')).toBeInTheDocument();
    });
  });

  it('table with auto pagination should work', async () => {
    const { getByTestId, getByRole, queryByText, rerender } = renderWithClient(
      <BasicTable
        dataLength={100}
        tableProps={{
          enablePagination: true,
          pageSize: 20,
        }}
      />,
    );

    expect(getByTestId('pagination-container')).toBeInTheDocument();
    expect(getByRole('button', { name: /1/i })).toBeInTheDocument();
    expect(getByRole('button', { name: /5/i })).toBeInTheDocument();
    expect(getByRole('button', { name: /next/i })).toBeInTheDocument();
    expect(getByRole('button', { name: /previous/i })).toBeInTheDocument();
    expect(getByRole('cell', { name: /fruit 0/i })).toBeInTheDocument();

    fireEvent.click(getByRole('button', { name: /2/i }));

    await waitFor(() => {
      expect(queryByText('fruit 0')).not.toBeInTheDocument();
      expect(getByRole('cell', { name: /fruit 20/i })).toBeInTheDocument();
    });

    rerender(
      <BasicTable
        dataLength={100}
        tableProps={{
          enablePagination: true,
          pageSize: 10,
        }}
      />,
    );

    await waitFor(() => {
      expect(getByRole('button', { name: /10/i })).toBeInTheDocument();
    });
  });
});
