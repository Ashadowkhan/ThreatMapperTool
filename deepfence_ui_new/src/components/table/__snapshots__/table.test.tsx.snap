// Vitest Snapshot v1

exports[`Component Table > should render a basic table > cells with border 0 1`] = `"<td style=\\"width: 150px;\\" class=\\"p-4 text-sm text-gray-900 dark:text-white font-normal border-b border-gray-200 dark:border-gray-700\\">0</td>"`;

exports[`Component Table > should render a basic table > cells with border 1 1`] = `"<td style=\\"width: 150px;\\" class=\\"p-4 text-sm text-gray-900 dark:text-white font-normal border-b border-gray-200 dark:border-gray-700\\">Fruit 0</td>"`;

exports[`Component Table > should render a basic table > cells with border 2 1`] = `"<td style=\\"width: 150px;\\" class=\\"p-4 text-sm text-gray-900 dark:text-white font-normal border-b border-gray-200 dark:border-gray-700\\">Description for 0</td>"`;

exports[`Component Table > should render a basic table > cells without border 3 1`] = `"<td style=\\"width: 150px;\\" class=\\"p-4 text-sm text-gray-900 dark:text-white font-normal\\">1</td>"`;

exports[`Component Table > should render a basic table > cells without border 4 1`] = `"<td style=\\"width: 150px;\\" class=\\"p-4 text-sm text-gray-900 dark:text-white font-normal\\">Fruit 1</td>"`;

exports[`Component Table > should render a basic table > cells without border 5 1`] = `"<td style=\\"width: 150px;\\" class=\\"p-4 text-sm text-gray-900 dark:text-white font-normal\\">Description for 1</td>"`;

exports[`Component Table > should render a basic table > row with hover state on normal table 1`] = `
<tr
  class="hover:!bg-gray-100 dark:hover:!bg-gray-600 transition-colors"
>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    1
  </td>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    Fruit 1
  </td>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    Description for 1
  </td>
</tr>
`;

exports[`Component Table > should render a striped table > cells should not have border 1`] = `
<td
  class="p-4 text-sm text-gray-900 dark:text-white font-normal"
  style="width: 150px;"
>
  Fruit 0
</td>
`;

exports[`Component Table > should render a striped table > even rows should have darker background 1`] = `
<tr
  class="bg-gray-50 dark:bg-gray-700 hover:!bg-gray-100 dark:hover:!bg-gray-600 transition-colors"
>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    1
  </td>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    Fruit 1
  </td>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    Description for 1
  </td>
</tr>
`;

exports[`Component Table > should render a striped table > odd rows should have lighter background 1`] = `
<tr
  class="hover:!bg-gray-100 dark:hover:!bg-gray-600 transition-colors"
>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    0
  </td>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    Fruit 0
  </td>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    Description for 0
  </td>
</tr>
`;

exports[`Component Table > should render a striped table > row with hover state on striped table 1`] = `
<tr
  class="bg-gray-50 dark:bg-gray-700 hover:!bg-gray-100 dark:hover:!bg-gray-600 transition-colors"
>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    1
  </td>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    Fruit 1
  </td>
  <td
    class="p-4 text-sm text-gray-900 dark:text-white font-normal"
    style="width: 150px;"
  >
    Description for 1
  </td>
</tr>
`;
